### ISSUE TITLE
#include with long name are ignored

### ISSUE THREAD

**Issue by rt-2** [Southclaws/pawn#123]
When I try
`#include "../include/functions/func_movePlayerGently.pwn"`
it works but when I do:
`#include "../include/functions/saarp_func_movePlayerGently.pwn"`
It doesn't.

Thank you

**Comment by Y-Less** [comment]
`\`, not `/`!

**Comment by Y-Less** [comment]
Also, full details of what the compiler is doing:

When you include a file, you get a symbol which is `_inc_FILENAME` defined.  The exact part of the path used for this symbol is complicated, but if you use the correct directory separator it will just be the file name (if you look in YSI, there is a combination of both `/` and `\` used in various places for exactly this reason).  For `a_samp` this is exactly the same as doing:

```pawn
#define _inc_a_samp
```

And is how the compiler determines which files have been included before for its internal include guards (enabled in the updated compiler by `-Z`).

Symbols in PAWN have a maximum length of 32 characters, and this is difficult to change because the same limit exists in both the compiler and VM (though again, the rules on which symbol lengths could be increased and which can't are complex).  So files with names that are too long will generate symbols that are too long, and they will get truncated (don't forget the extra five characters for `_inc_`).  Because of the truncation and the selection of path segments to use for the symbol generation, it is possible for two different files to generate the same symbol, and thus for the include guard to think that the file has already been included.

Long story short - why do you have separate files for separate functions (there's a file limit too)?

**Comment by Y-Less** [comment]
Because I'm a completionist, the path segment used for generating the symbol name is the last part after the native directory separator (`\` in Windows).  If you include a file with:

```pawn
#include "some\dir\file.pwn"
```

The part after the final `\` (excluding extensions) is `file`, so the generated include guard symbol is `_inc_file`.  As a result, these lines will not work as you may expect:

```pawn
#include "some\dir\file.pwn"
#include "other\dir\file.pwn"
```

The first line generates `_inc_file`, the second line looks for `_inc_file`, finds it, and thus thinks that that file has already been included, so the second file will NOT get included.

If you use a mixture of directory separators, the rule still applies - the symbol is everything after the last native directory separator:

```pawn
#include "some\dir/file.pwn"
```

Generates the symbol `_inc_dir/file`.  Yes, exactly that.  A side-effect of this is that you can't possibly remove the symbols:

```pawn
#include "some\dir/file.pwn"
#undef _inc_dir/file
```

That is not valid code and errors because `/` is not a valid symbol character, it is introduced to that symbol through a side-channel and can't be used in text.  Another side-effect of this include method is that the present working directory is not updated correctly.  Assuming that `some\dir/file.pwn` includes this line:

```pawn
#include "other.pwn"
```

You might expect that to include the file `some\dir/other.pwn`, since `""` is for relative inclusion.  However, because the working directory is everything before the final native separator, the pwd is just `some`, and only the file `some\other.pwn` gets included.  Because it is only the final native separator that counts for anything, this works normally:

```pawn
#include "some/dir\file.pwn"
```

Many files in YSI have the name `impl.inc`, as the core implementation of a library.  These are all included via something like:

```pawn
#include "y_commands/impl.inc"
```

As y_commands is included by:

```pawn
#include <YSI_Visual\y_commands>
```

The directory separators become transitive, and so the first line is equivalent to:

```pawn
#include <YSI_Visual\y_commands/impl.inc>
```

Hence the combination of separators remains.  This is why one of the most common bugs is people doing:

```pawn
#include <YSI/y_ini>
```

And getting many strange errors.  That makes the complete inclusion path:

```pawn
#include <YSI/y_ini/impl.inc>
```

There is no native directory separator there, and the rules change.  I'm not even sure of the exact semantics when every separator is wrong.

**Comment by Y-Less** [comment]
As for which symbol lengths can be increased, the actual answer is "any not `public`".  The VM, when searching for callbacks and timers, looks through the AMX header for string representations of function names, paired with pointers to those functions.  After compilation, this is the only place that the human-readable versions of the function names are kept, and only for those marked `public` or `@`.  Plain functions and those marked `stock` are compiled to addresses only.  Thus public functions are always limited to the maximum length of function name that the VM looks at, which is currently 32 (though I might check this because I don't know for certain that the length is truly used in that comparison even if it is defined).  Others can be increased arbitrarilly in the compiler, as their names will just vanish.  Of course, this will cause backwards-compatability issues for anyone writing libraries or other code that will be used by people on either compiler.

**Comment by Y-Less** [comment]
https://github.com/Zeex/pawn/blob/1697d6f8b57822710441d5116d80199102eb9c41/source/amx/amx.c#L1467-L1492

That's the function in the VM responsible for finding public functions by name, and it does use the max symbol length here:

https://github.com/Zeex/pawn/blob/1697d6f8b57822710441d5116d80199102eb9c41/source/amx/amx.c#L1470

**Comment by rt-2** [comment]
In response to your first 2 comments:
I thought that the new compiler removed the include guard?
`#include "..\include\functions\saarp_func_movePlayerGently.pwn"`
Worked for me!

Thank you very much!

**Comment by Y-Less** [comment]
Yes, though `-Z` puts it back (if that applies to you), and I don't know if the symbol generation is disabled, or just the checks.

**Comment by rt-2** [comment]
Unfortunatly I have no parameters, isn't werird that it is still doing the 'include guard' thing?
![image](https://cloud.githubusercontent.com/assets/3452428/21018950/e181d9b8-bd3c-11e6-9aca-721ff743012c.png)

**Comment by Y-Less** [comment]
I don't even know if those parameters are used even when specified.  I know that there are some default ones (`-;+ -(+` for example) and the rest are pulled from `pawn.cfg`.  But honestly, I don't know if it should be doing the guard or not, why it might be, or even if that was the actual problem and not another one stemming from using `/`.

**Comment by rt-2** [comment]
In fact I think you were completely right, because there was some very strange behavior, that later I discovered that the second file with the \ vs / problem didn't get included, Probably because of the reason you mentioned about the name and the include guard.

The reason I think there should be no Include guard is because of the 'No automatic #include guards' section at https://github.com/Zeex/pawn/wiki/What's-new#no-automatic-include-guards (near the end).
SO maybe I read wrong, or maybe the documentation is wrong, but I thought it should be off.

Also, I haven't find any 'pawn.cfg' file in the pawno directory...

Thanks again for your help.
Again, my problem has been fixed with the \ instead of /. I'm just trying to get to the bottom of this, I am curious and a perfectionist.

**Comment by Y-Less** [comment]
Same, I'm not sure what the problem is either - could even be something like the filename ending up too long in some check and thus getting dropped (in which case your original thought that it was a too long filename was semi right, once the path was included in that length as well).  But honestly, I don't know - would have to consult the compiler source for that.

**Comment by rt-2** [comment]
I'll make a bug, I'll write "Compiler 'No automatic #include guards' not working" and refer to this issue, because I think there is seriously something wrong with this issue vs the documentation that, I think, is worth communicating...

Thank you VERY much for your help!

**Comment by spacemud** [comment]
> Long story short - why do you have separate files for separate functions (there's a file limit too)?

Just curious - what is that file limit?

**Comment by Y-Less** [comment]
Not sure, I think it is lower on Linux, but both are limited by the maximum number of file handles allowed per process on an OS (this can be adjusted, but IIRC you have to recompile the kernel to do so).

**Comment by rt-2** [comment]
But is it in the 256 range or the 2,147,483,647 range?
Is there somewhere I can know what is the number?
I am starting to separate my game-mode into files per script, and I wanna know now if 'm shooting myself in the foot.
Thanks

**Comment by Y-Less** [comment]
I have no idea, but YSI is huge, and includes some files many multiple times, so I've never seen anyone actually legitimately hit it.

**Comment by YashasSamaga** [comment]
The compiler closes the includes after parsing. Therefore, I think there shouldn't be any file limit. 

A situation like this might cause problems.
a.pwn includes b.inc
b.inc includes c.inc
c.inc includes d.inc
.
.
.
The number of files open simultaneously is limited to FOPEN_MAX (which is of course implementation dependent).
